### PR TITLE
CHE-102 - Idle detection of che-server and workspaces

### PR DESCRIPTION
**Pull request by snjeza**
### What does this PR do?

Detect when a workspace and che-server are idle

### What issues does this PR fix or reference?

https://issues.jboss.org/browse/CHE-102

### CODE REVIEW

File: wsagent/wsagent-local/src/main/java/org/eclipse/che/idle/WorkspaceIdleFilter.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    public void init(FilterConfig filterConfig) throws ServletException {
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        if (!stopped.get()) {
+            future.cancel(true);
+            future = executor.schedule(this::sendEvent, timeout, TimeUnit.MILLISECONDS);
+            LOG.info("Idling workspace " + WorkspaceIdProvider.getWorkspaceId() + " scheduled [timeout="
```
Comment (l0rd) [inline wsagent/wsagent-local/src/main/java/org/eclipse/che/idle/WorkspaceIdleFilter.java]: Writing on disk at every HTTP request is bad for performances. Please remove this log.

File: wsagent/wsagent-local/src/main/java/org/eclipse/che/idle/WorkspaceIdleFilter.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        }
+        chain.doFilter(request, response);
+    }
+
+    private void sendEvent() {
+        try {
+            httpJsonRequestFactory.fromUrl(apiEndpoint + "/workspace/" + WorkspaceIdProvider.getWorkspaceId() + "/runtime")
+                                  .useDeleteMethod()
+                                  .request();
+            LOG.info("Workspace " + WorkspaceIdProvider.getWorkspaceId() + " has been stopped.");
```
Comment (l0rd) [inline wsagent/wsagent-local/src/main/java/org/eclipse/che/idle/WorkspaceIdleFilter.java]: Please use a more explicit message here: `"Idling workspace " + WorkspaceIdProvider.getWorkspaceId() + " after " + timeout / 1000  + " seconds of inactivity."`

**Review by l0rd — APPROVED** [review]
Great job @snjeza. I've added some comments but since these are not critical, if you are ok, we can merge this PR and you can address the remaining issues in a successive PR. What do you think?

**Comment by l0rd** [comment]
But you still need to rebase though...

**Comment by l0rd** [comment]
@snjeza can you rebase this PR so we merge it? 

We may decide to revert cd9552e18fbf35b490bf48152f2fccb6aa8f0c3a but that can take a few days so I would like to merge that first.

**Comment by snjeza** [comment]
I will do it, but the current openshift branch is broken. See https://issues.jboss.org/browse/CHE-180?focusedCommentId=13393998&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13393998.

**Comment by TylerJewell** [comment]
Added `target/branch` label.  Please apply this label to all non-master branch PRs going forward, please.

**Comment by l0rd** [comment]
@snjeza CHE-180 and this issue aren't related

Il 17 apr 2017 4:08 PM, "Snjeza" <notifications@github.com> ha scritto:

File: wsagent/wsagent-local/src/main/java/org/eclipse/che/idle/WorkspaceIdleFilter.java (hunk 2 above)
Comment (snjeza) [reply wsagent/wsagent-local/src/main/java/org/eclipse/che/idle/WorkspaceIdleFilter.java]: We don't idle a workspace. We stop it. 
I have used: "Stopping workspace " + WorkspaceIdProvider.getWorkspaceId() + " after " + timeout / 1000 + " seconds of inactivity."

File: plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftConnector.java [hunk 3]
```diff
@@ -159,6 +172,12 @@
     private static final String OPENSHIFT_VOLUME_STORAGE_CLASS           = "volume.beta.kubernetes.io/storage-class";
     private static final String OPENSHIFT_VOLUME_STORAGE_CLASS_NAME      = "che-workspace";
     private static final String OPENSHIFT_IMAGE_PULL_POLICY_IFNOTPRESENT = "IfNotPresent";
+    private static final String IDLING_ALPHA_OPENSHIFT_IO_IDLED_AT       = "idling.alpha.openshift.io/idled-at";
+    private static final String IDLING_ALPHA_OPENSHIFT_IO_PREVIOUS_SCALE = "idling.alpha.openshift.io/previous-scale";
+    private static final String OPENSHIFT_CHE_SERVER_DEPLOYMENT_NAME     = "che";
+    private static final String OPENSHIFT_CHE_SERVER_SERVICE_NAME        = "che-host";
```
[the remark this answers is not in the document]
Comment (snjeza) [reply plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftConnector.java:178]: We need the deployment config and service of the che server (che and che-host), not of a workspace. How can I get them?

[the remark this answers is not in the document]
Comment (l0rd) [reply plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftConnector.java:178]: You are right. We need to think about that.

File: wsagent/wsagent-local/src/main/java/org/eclipse/che/idle/WorkspaceIdleFilter.java (hunk 2 above)
Comment (l0rd) [reply wsagent/wsagent-local/src/main/java/org/eclipse/che/idle/WorkspaceIdleFilter.java]: That's perfect!

**Pull request merged**